src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,180 @@
+        if (lastNode == null) {
+            firstNode = newNode;
+        } else {
+            lastNode.next = newNode;
+        }
+        lastNode = newNode;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index == size) {
+            add(value);
+        } else {
+            checkRangeForAdd(index);
+            Node<T> currentNode = getNode(index);
+            Node<T> newNode = new Node<>(currentNode.prev, value, currentNode);
+            Node<T> prev = currentNode.prev;
+            newNode.prev = prev;
+            newNode.next = currentNode;
+            currentNode.prev = newNode;
+            if (prev == null) {
+                firstNode = newNode;
+            } else {
+                prev.next = newNode;
+            }
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkRangeForAdd(index);
+        return getNode(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkRangeForAdd(index);
+        Node<T> currentNode = getNode(index);
+        T oldValue = currentNode.value;
+        currentNode.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkRangeForAdd(index);
+        Node<T> currentNode = getNode(index);
+        if (currentNode.prev == null) {
+            firstNode = currentNode.next;
+        } else {
+            currentNode.prev.next = currentNode.next;
+        }
+        if (currentNode.next == null) {
+            lastNode = currentNode.prev;
+        } else {
+            currentNode.next.prev = currentNode.prev;
+        }
+        size--;
+        return currentNode.value;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> currentNode = firstNode;
+        while (currentNode != null) {
+            if (elementsAreEqual(object, currentNode)) {
+                if (currentNode.prev == null) {
+                    firstNode = currentNode.next;
+                } else {
+                    currentNode.prev.next = currentNode.next;
+                }
+                if (currentNode.next == null) {
+                    lastNode = currentNode.prev;
+                } else {
+                    currentNode.next.prev = currentNode.prev;
+                }
+                size--;
+                return true;
+            }
+            currentNode = currentNode.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void checkRangeForAdd(int index) {
```

Comment:
I think, that  u need to rename this method cause u don't u this check only for "add"  methods.
```suggestion
    private void checkRange(int index) {
```